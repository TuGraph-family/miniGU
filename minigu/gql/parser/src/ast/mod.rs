pub mod catalog;
pub mod common;
pub mod data;
pub mod lexical;
pub mod object_expr;
pub mod object_ref;
pub mod predicate;
pub mod procedure_call;
pub mod procedure_spec;
pub mod program;
pub mod query;
pub mod session;
pub mod transaction;
pub mod type_element;
pub mod value_expr;
pub mod variable;

pub use catalog::*;
pub use common::*;
pub use data::*;
pub use lexical::*;
pub use object_expr::*;
pub use object_ref::*;
pub use predicate::*;
pub use procedure_call::*;
pub use procedure_spec::*;
pub use program::*;
pub use query::*;
pub use session::*;
pub use transaction::*;
pub use type_element::*;
pub use value_expr::*;
pub use variable::*;
