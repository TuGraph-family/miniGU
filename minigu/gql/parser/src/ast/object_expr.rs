//! AST definitions for *object expressions*.

use super::{Expr, GraphRef, Ident};
use crate::macros::base;
use crate::span::Spanned;

#[apply(base)]
pub enum GraphExpr {
    Name(Ident),
    Ref(GraphRef),
    Object(ObjectExpr),
    Current,
}

#[apply(base)]
pub enum ObjectExpr {
    Variable(Spanned<Expr>),
    Expr(Spanned<Expr>),
}
