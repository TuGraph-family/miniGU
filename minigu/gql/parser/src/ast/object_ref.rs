//! AST definitions for *object references*.

use super::Ident;
use crate::macros::base;
use crate::span::{OptSpanned, VecSpanned};

#[apply(base)]
pub enum PredefinedSchemaRef {
    Home,
    Current,
}

#[apply(base)]
pub enum SchemaRef {
    Absolute(SchemaPath),
    Relative(SchemaPath),
    Predefined(PredefinedSchemaRef),
    Parameter(Ident),
}

pub type SchemaPath = VecSpanned<SchemaPathSegment>;

#[apply(base)]
pub enum SchemaPathSegment {
    /// A named object (schema or directory), e.g., `a`.
    Name(Ident),
    /// Parent directory, e.g., `..`.
    Parent,
}

#[apply(base)]
pub enum GraphRef {
    Name(Ident),
    Parameter(Ident),
    Ref(CatalogObjectRef),
    Home,
}

#[apply(base)]
pub enum ProcedureRef {
    Ref(CatalogObjectRef),
    Parameter(Ident),
}

#[apply(base)]
pub enum GraphTypeRef {
    Ref(CatalogObjectRef),
    Parameter(Ident),
}

#[apply(base)]
pub enum BindingTableRef {
    Name(Ident),
    Ref(CatalogObjectRef),
    Parameter(Ident),
}

#[apply(base)]
pub struct CatalogObjectRef {
    pub schema: OptSpanned<SchemaRef>,
    pub objects: VecSpanned<Ident>,
}
