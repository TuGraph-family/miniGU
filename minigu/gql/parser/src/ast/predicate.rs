//! AST definitions for *predicates*.

use super::{Expr, Ident, LabelExpr, MatchStatement, Procedure, ValueType};
use crate::macros::base;
use crate::span::{BoxSpanned, Spanned, VecSpanned};

#[apply(base)]
pub enum Predicate {
    Exists(Exists),
    Null {
        expr: BoxSpanned<Expr>,
        not: bool,
    },
    Typed {
        expr: BoxSpanned<Expr>,
        value_type: BoxSpanned<ValueType>,
        not: bool,
    },
    Directed {
        edge: Spanned<Ident>,
        not: bool,
    },
    Labeled {
        element: Spanned<Ident>,
        label: Spanned<LabelExpr>,
    },
    SrcOf {
        node: Spanned<Ident>,
        edge: Spanned<Ident>,
        not: bool,
    },
    DstOf {
        node: Spanned<Ident>,
        edge: Spanned<Ident>,
        not: bool,
    },
    AllDifferent(VecSpanned<Ident>),
    Same(VecSpanned<Ident>),
    PropertyExists {
        element: Spanned<Ident>,
        property: Spanned<Ident>,
    },
}

#[apply(base)]
pub enum Exists {
    Pattern,
    Match(VecSpanned<MatchStatement>),
    Nested(BoxSpanned<Procedure>),
}
