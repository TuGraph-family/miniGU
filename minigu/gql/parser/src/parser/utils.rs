use winnow::Parser;
use winnow::stream::{Location, Stream};

use crate::span::Spanned;

/// A helper macro for executing a parser on an input string. **This should only be used in unit
/// tests!**
#[cfg(all(test, feature = "serde"))]
macro_rules! parse {
    ($parser:expr, $input:literal) => {{
        let input = $input;
        $crate::parser::token::tokenize(input)
            .ok()
            .and_then(|tokens| {
                let stream = crate::parser::token::build_token_stream(&tokens, Default::default());
                $parser.parse(stream).ok()
            })
    }};
}

#[cfg(all(test, feature = "serde"))]
pub(super) use parse;

/// A helper trait for wrapping the output in [`Spanned`].
pub(super) trait ToSpanned<I, O, E>: Parser<I, O, E> {
    /// Wraps the output in [`Spanned`].
    #[inline(always)]
    fn spanned(self) -> impl Parser<I, Spanned<O>, E>
    where
        Self: Sized,
        I: Stream + Location,
    {
        self.with_span().map(|(inner, span)| Spanned(inner, span))
    }
}

/// A helper trait for operating on the value wrapped in [`Spanned`].
pub(super) trait SpannedParserExt<I, O, E>: Parser<I, Spanned<O>, E> {
    /// Unwraps the output from [`Spanned`].
    #[inline(always)]
    fn unspanned(self) -> impl Parser<I, O, E>
    where
        Self: Sized,
    {
        self.map(|spanned| spanned.0)
    }

    /// Maps the inner value of the output while keeping the span unchanged.
    #[inline(always)]
    fn map_inner<F, O2>(self, mut f: F) -> impl Parser<I, Spanned<O2>, E>
    where
        Self: Sized,
        F: FnMut(O) -> O2,
    {
        self.map(move |Spanned(inner, span)| Spanned(f(inner), span))
    }

    /// Updates the span of the output while keeping the inner value unchanged.
    #[inline(always)]
    fn update_span(self) -> impl Parser<I, Spanned<O>, E>
    where
        Self: Sized,
        I: Stream + Location,
    {
        self.with_span().map(|(mut output, new_span)| {
            output.1 = new_span;
            output
        })
    }
}

impl<I, O, E, P> ToSpanned<I, O, E> for P where P: Parser<I, O, E> {}

impl<I, O, E, P> SpannedParserExt<I, O, E> for P where P: Parser<I, Spanned<O>, E> {}

/// A helper macro for defining parser aliases.
macro_rules! def_parser_alias {
    ($name:ident, $parser:ident, $output:ty) => {
        #[inline(always)]
        pub fn $name(
            input: &mut $crate::parser::token::TokenStream,
        ) -> winnow::ModalResult<$output> {
            $parser(input)
        }
    };
}

pub(super) use def_parser_alias;
