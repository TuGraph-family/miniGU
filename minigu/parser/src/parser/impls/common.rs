use winnow::combinator::{
    alt, delimited, dispatch, empty, fail, opt, preceded, repeat, separated, separated_pair, seq,
    terminated,
};
use winnow::token::{any, one_of};
use winnow::{ModalResult, Parser};

use super::lexical::{
    binding_variable, edge_synonym, edges_synonym, element_variable, field_name, label_name,
    path_variable, property_name, subpath_variable, unsigned_integer,
};
use super::object_expr::graph_expression;
use super::object_ref::schema_reference;
use super::predicate::search_condition;
use super::value_expr::{
    aggregating_value_expression, binding_variable_reference, value_expression,
};
use crate::ast::*;
use crate::imports::Box;
use crate::lexer::TokenKind;
use crate::parser::impls::value_expr::non_negative_integer_specification;
use crate::parser::precedence::{Assoc, Precedence, precedence};
use crate::parser::token::{Token, TokenStream};
use crate::parser::utils::{
    SpannedParserExt, ToSpanned, def_parser_alias, peek1, peek2, peek3, take1, take2,
};
use crate::span::{Spanned, VecSpanned};

const PREC_INIT: Precedence = 0;
const PREC_DISJUNCTION: Precedence = 1;
const PREC_CONJUNCTION: Precedence = 2;
const PREC_NEGATION: Precedence = 3;

pub fn use_graph_clause(input: &mut TokenStream) -> ModalResult<Spanned<GraphExpr>> {
    preceded(TokenKind::Use, graph_expression).parse_next(input)
}

pub fn at_schema_clause(input: &mut TokenStream) -> ModalResult<Spanned<SchemaRef>> {
    preceded(TokenKind::At, schema_reference).parse_next(input)
}

pub fn yield_clause(input: &mut TokenStream) -> ModalResult<Spanned<Yield>> {
    preceded(
        TokenKind::Yield,
        separated(1.., yield_item, TokenKind::Comma),
    )
    .spanned()
    .parse_next(input)
}

pub fn yield_item(input: &mut TokenStream) -> ModalResult<Spanned<YieldItem>> {
    seq! {YieldItem {
        name: yield_item_name,
        alias: opt(yield_item_alias),
    }}
    .spanned()
    .parse_next(input)
}

pub fn yield_item_alias(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    preceded(TokenKind::As, binding_variable).parse_next(input)
}

def_parser_alias!(yield_item_name, field_name, Spanned<Ident>);

pub fn is_or_colon(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Is, TokenKind::Colon))
        .void()
        .parse_next(input)
}

pub fn graph_pattern_binding_table(
    input: &mut TokenStream,
) -> ModalResult<Spanned<GraphPatternBindingTable>> {
    seq! {GraphPatternBindingTable {
        pattern: graph_pattern,
        yield_clause: opt(graph_pattern_yield_clause).map(Option::unwrap_or_default),
    }}
    .spanned()
    .parse_next(input)
}

pub fn graph_pattern_yield_clause(input: &mut TokenStream) -> ModalResult<VecSpanned<Ident>> {
    preceded(
        TokenKind::Yield,
        separated(1.., binding_variable_reference, TokenKind::Comma),
    )
    .parse_next(input)
}

pub fn graph_pattern(input: &mut TokenStream) -> ModalResult<Spanned<GraphPattern>> {
    seq! {GraphPattern {
        match_mode: opt(match_mode),
        patterns: separated(1.., path_pattern, TokenKind::Comma),
        keep: opt(keep_clause),
        where_clause: opt(graph_pattern_where_clause),
    }}
    .spanned()
    .parse_next(input)
}

pub fn keep_clause(input: &mut TokenStream) -> ModalResult<Spanned<PathPatternPrefix>> {
    preceded(TokenKind::Keep, path_pattern_prefix).parse_next(input)
}

pub fn match_mode(input: &mut TokenStream) -> ModalResult<Spanned<MatchMode>> {
    dispatch! {take1;
        TokenKind::Repeatable => element_bindings_or_elements.value(MatchMode::Repeatable),
        TokenKind::Different => edge_bindings_or_edges.value(MatchMode::Different),
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn element_bindings_or_elements(input: &mut TokenStream) -> ModalResult<()> {
    alt((
        (TokenKind::Element, opt(TokenKind::Bindings)).void(),
        TokenKind::Elements.void(),
    ))
    .parse_next(input)
}

pub fn edge_bindings_or_edges(input: &mut TokenStream) -> ModalResult<()> {
    alt((
        (edge_synonym, opt(TokenKind::Bindings)).void(),
        edges_synonym,
    ))
    .parse_next(input)
}

pub fn element_pattern(input: &mut TokenStream) -> ModalResult<Spanned<ElementPattern>> {
    dispatch! {peek1;
        TokenKind::LeftParen => node_pattern,
        kind if kind.is_prefix_of_edge_pattern() => edge_pattern,
        _ => fail
    }
    .parse_next(input)
}

pub fn node_pattern(input: &mut TokenStream) -> ModalResult<Spanned<ElementPattern>> {
    delimited(
        TokenKind::LeftParen,
        element_pattern_filler,
        TokenKind::RightParen,
    )
    .map_inner(ElementPattern::Node)
    .update_span()
    .parse_next(input)
}

pub fn edge_pattern(input: &mut TokenStream) -> ModalResult<Spanned<ElementPattern>> {
    dispatch! {peek1;
        kind if kind.is_prefix_of_full_edge_pattern() => full_edge_pattern,
        kind if kind.is_prefix_of_abbreviated_edge_pattern() => {
            abbreviated_edge_pattern
                .map(|kind| ElementPattern::Edge {
                    kind,
                    filler: ElementPatternFiller {
                        variable: None,
                        label: None,
                        predicate: None,
                    },
                })
                .spanned()
        },
        _ => fail
    }
    .parse_next(input)
}

pub fn full_edge_pattern(input: &mut TokenStream) -> ModalResult<Spanned<ElementPattern>> {
    // This avoids matching `element_pattern_filler` multiple times.
    (
        any.map(|t: &Token| &t.kind),
        element_pattern_filler.unspanned(),
        any.map(|t: &Token| &t.kind),
    )
        .verify_map(|(left, filler, right)| {
            let kind = match (left, right) {
                (TokenKind::LeftArrowBracket, TokenKind::RightBracketMinus) => {
                    EdgePatternKind::Left
                }
                (TokenKind::TildeLeftBracket, TokenKind::RightBracketTilde) => {
                    EdgePatternKind::Undirected
                }
                (TokenKind::MinusLeftBracket, TokenKind::BracketRightArrow) => {
                    EdgePatternKind::Right
                }
                (TokenKind::LeftArrowTildeBracket, TokenKind::RightBracketTilde) => {
                    EdgePatternKind::LeftUndirected
                }
                (TokenKind::TildeLeftBracket, TokenKind::BracketTildeRightArrow) => {
                    EdgePatternKind::RightUndirected
                }
                (TokenKind::LeftArrowBracket, TokenKind::BracketRightArrow) => {
                    EdgePatternKind::LeftRight
                }
                (TokenKind::MinusLeftBracket, TokenKind::RightBracketMinus) => EdgePatternKind::Any,
                _ => return None,
            };
            Some(ElementPattern::Edge { kind, filler })
        })
        .spanned()
        .parse_next(input)
}

pub fn abbreviated_edge_pattern(input: &mut TokenStream) -> ModalResult<EdgePatternKind> {
    dispatch! {take1;
        TokenKind::LeftArrow => empty.value(EdgePatternKind::Left),
        TokenKind::Tilde => empty.value(EdgePatternKind::Undirected),
        TokenKind::RightArrow => empty.value(EdgePatternKind::Right),
        TokenKind::LeftArrowTilde => empty.value(EdgePatternKind::LeftUndirected),
        TokenKind::TildeRightArrow => empty.value(EdgePatternKind::RightUndirected),
        TokenKind::LeftMinusRight => empty.value(EdgePatternKind::LeftRight),
        TokenKind::Minus => empty.value(EdgePatternKind::Any),
        _ => fail
    }
    .parse_next(input)
}

pub fn element_pattern_filler(
    input: &mut TokenStream,
) -> ModalResult<Spanned<ElementPatternFiller>> {
    seq! {ElementPatternFiller {
        variable: opt(element_variable_declaration),
        label: opt(is_label_expression),
        predicate: opt(element_pattern_predicate),
    }}
    .spanned()
    .parse_next(input)
}

pub fn element_pattern_predicate(
    input: &mut TokenStream,
) -> ModalResult<Spanned<ElementPatternPredicate>> {
    dispatch! {peek1;
        TokenKind::Where => {
            preceded(TokenKind::Where, search_condition)
                .map(ElementPatternPredicate::Where)
        },
        TokenKind::LeftBrace => {
            element_property_specification.map(ElementPatternPredicate::Property)
        },
        _ => fail
    }
    .spanned()
    .parse_next(input)
}

pub fn element_property_specification(
    input: &mut TokenStream,
) -> ModalResult<VecSpanned<FieldOrProperty>> {
    delimited(
        TokenKind::LeftBrace,
        separated(1.., property_key_value_pair, TokenKind::Comma),
        TokenKind::RightBrace,
    )
    .parse_next(input)
}

pub fn property_key_value_pair(input: &mut TokenStream) -> ModalResult<Spanned<FieldOrProperty>> {
    seq! {FieldOrProperty {
        name: property_name,
        _: TokenKind::Colon,
        value: value_expression,
    }}
    .spanned()
    .parse_next(input)
}

pub fn offset_synonym(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Offset, TokenKind::Skip))
        .void()
        .parse_next(input)
}

pub fn is_label_expression(input: &mut TokenStream) -> ModalResult<Spanned<LabelExpr>> {
    preceded(is_or_colon, label_expression).parse_next(input)
}

def_parser_alias!(
    element_variable_declaration,
    element_variable,
    Spanned<Ident>
);

#[derive(Clone)]
enum InfixLabelOp {
    Conjunction,
    Disjunction,
}

fn label_expression_primary(input: &mut TokenStream) -> ModalResult<Spanned<LabelExpr>> {
    dispatch! {peek1;
        TokenKind::LeftParen => {
            delimited(TokenKind::LeftParen, label_expression, TokenKind::RightParen).update_span()
        },
        TokenKind::Percent => TokenKind::Percent.value(LabelExpr::Wildcard).spanned(),
        _ => label_name.map_inner(LabelExpr::Label),
    }
    .parse_next(input)
}

fn label_expression_prefix(input: &mut TokenStream) -> ModalResult<(Precedence, Spanned<()>)> {
    TokenKind::Exclamation
        .void()
        .spanned()
        .map(|span| (PREC_NEGATION, span))
        .parse_next(input)
}

fn label_expression_infix(
    input: &mut TokenStream,
) -> ModalResult<(Assoc, Precedence, InfixLabelOp)> {
    dispatch! {take1;
        TokenKind::Ampersand => empty.value((Assoc::Left, PREC_CONJUNCTION, InfixLabelOp::Conjunction)),
        TokenKind::VerticalBar => empty.value((Assoc::Left, PREC_DISJUNCTION, InfixLabelOp::Disjunction)),
        _ => fail
    }
    .parse_next(input)
}

pub fn label_expression(input: &mut TokenStream) -> ModalResult<Spanned<LabelExpr>> {
    precedence(
        PREC_INIT,
        label_expression_primary,
        label_expression_prefix,
        fail,
        label_expression_infix,
        |op, a| {
            let span = op.1.start..a.1.end;
            let inner = LabelExpr::Negation(Box::new(a));
            Ok(Spanned(inner, span))
        },
        |_, ()| unreachable!(),
        |a, op, b| {
            let span = a.1.start..b.1.end;
            let inner = match op {
                InfixLabelOp::Conjunction => LabelExpr::Conjunction(Box::new(a), Box::new(b)),
                InfixLabelOp::Disjunction => LabelExpr::Disjunction(Box::new(a), Box::new(b)),
            };
            Ok(Spanned(inner, span))
        },
    )
    .parse_next(input)
}

pub fn graph_pattern_quantifier(
    input: &mut TokenStream,
) -> ModalResult<Spanned<PatternQuantifier>> {
    let mut fixed_or_general = dispatch! {peek3;
        (TokenKind::LeftBrace, kind, TokenKind::RightBrace)
            if kind != &TokenKind::Comma =>
        {
            fixed_quantifier
        },
        _ => general_quantifier,
    };
    dispatch! {peek1;
        TokenKind::Asterisk => {
            TokenKind::Asterisk
                .value(PatternQuantifier::Asterisk)
                .spanned()
        },
        TokenKind::Plus => {
            TokenKind::Plus
                .value(PatternQuantifier::Plus)
                .spanned()
        },
        TokenKind::LeftBrace => fixed_or_general,
        _ => fail
    }
    .parse_next(input)
}

pub fn fixed_quantifier(input: &mut TokenStream) -> ModalResult<Spanned<PatternQuantifier>> {
    delimited(
        TokenKind::LeftBrace,
        unsigned_integer,
        TokenKind::RightBrace,
    )
    .map(PatternQuantifier::Fixed)
    .spanned()
    .parse_next(input)
}

pub fn general_quantifier(input: &mut TokenStream) -> ModalResult<Spanned<PatternQuantifier>> {
    delimited(
        TokenKind::LeftBrace,
        separated_pair(
            opt(unsigned_integer),
            TokenKind::Comma,
            opt(unsigned_integer),
        ),
        TokenKind::RightBrace,
    )
    .map(|(lower_bound, upper_bound)| PatternQuantifier::General {
        lower_bound,
        upper_bound,
    })
    .spanned()
    .parse_next(input)
}

pub fn path_pattern(input: &mut TokenStream) -> ModalResult<Spanned<PathPattern>> {
    seq! {PathPattern {
        variable: opt(path_variable_declaration),
        prefix: opt(path_pattern_prefix),
        expr: path_pattern_expression,
    }}
    .spanned()
    .parse_next(input)
}

pub fn path_pattern_expression(input: &mut TokenStream) -> ModalResult<Spanned<PathPatternExpr>> {
    path_pattern_expression_inner.spanned().parse_next(input)
}

fn path_pattern_expression_inner(input: &mut TokenStream) -> ModalResult<PathPatternExpr> {
    let paths = path_term.parse_next(input)?;
    match opt(peek1).parse_next(input)? {
        Some(TokenKind::Alternation) => {
            let paths = repeat(1.., preceded(TokenKind::Alternation, path_term))
                .fold(
                    // FIXME: Remove `.clone()` here once `fold`'s init is changed to FnOnce.
                    // https://github.com/winnow-rs/winnow/issues/513
                    || [paths.clone()].into(),
                    |mut current: VecSpanned<_>, next| {
                        current.push(next);
                        current
                    },
                )
                .parse_next(input)?;
            Ok(PathPatternExpr::Alternation(paths))
        }
        Some(TokenKind::VerticalBar) => {
            let paths = repeat(1.., preceded(TokenKind::VerticalBar, path_term))
                .fold(
                    // FIXME: Same as above.
                    || [paths.clone()].into(),
                    |mut current: VecSpanned<_>, next| {
                        current.push(next);
                        current
                    },
                )
                .parse_next(input)?;
            Ok(PathPatternExpr::Union(paths))
        }
        _ => Ok(paths.0),
    }
}

pub fn path_term(input: &mut TokenStream) -> ModalResult<Spanned<PathPatternExpr>> {
    repeat(1.., path_factor)
        .map(PathPatternExpr::Concat)
        .spanned()
        .parse_next(input)
}

pub fn path_factor(input: &mut TokenStream) -> ModalResult<Spanned<PathPatternExpr>> {
    path_factor_inner.spanned().parse_next(input)
}

fn path_factor_inner(input: &mut TokenStream) -> ModalResult<PathPatternExpr> {
    let path = path_primary.parse_next(input)?;
    match opt(peek1).parse_next(input)? {
        Some(TokenKind::QuestionMark) => TokenKind::QuestionMark
            .value(PathPatternExpr::Optional(Box::new(path)))
            .parse_next(input),
        Some(TokenKind::Asterisk | TokenKind::Plus | TokenKind::LeftBrace) => {
            let quantifier = graph_pattern_quantifier(input)?;
            Ok(PathPatternExpr::Quantified {
                path: Box::new(path),
                quantifier,
            })
        }
        _ => Ok(path.0),
    }
}

// TODO: Add simplified path pattern expression.
pub fn path_primary(input: &mut TokenStream) -> ModalResult<Spanned<PathPatternExpr>> {
    dispatch! {peek1;
        TokenKind::LeftParen => alt((
            parenthesized_path_pattern_expression.map_inner(PathPatternExpr::Grouped),
            element_pattern.map_inner(PathPatternExpr::Pattern),
        )),
        kind if kind.is_prefix_of_edge_pattern() => {
            element_pattern.map_inner(PathPatternExpr::Pattern)
        },
        _ => fail,
    }
    .parse_next(input)
}

pub fn parenthesized_path_pattern_expression(
    input: &mut TokenStream,
) -> ModalResult<Spanned<GroupedPathPattern>> {
    seq! {GroupedPathPattern {
        _: TokenKind::LeftParen,
        variable: opt(subpath_variable_declaration),
        mode: opt(path_mode_prefix),
        expr: path_pattern_expression.map(Box::new),
        where_clause: opt(parenthesized_path_pattern_where_clause),
        _: TokenKind::RightParen,
    }}
    .spanned()
    .parse_next(input)
}

pub fn path_or_paths(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Path, TokenKind::Paths))
        .void()
        .parse_next(input)
}

pub fn path_pattern_prefix(input: &mut TokenStream) -> ModalResult<Spanned<PathPatternPrefix>> {
    dispatch! {peek1;
        TokenKind::Walk
        | TokenKind::Trail
        | TokenKind::Simple
        | TokenKind::Acyclic => path_mode_prefix.map_inner(PathPatternPrefix::PathMode),
        TokenKind::All
        | TokenKind::Any
        | TokenKind::Shortest => path_search_prefix.map_inner(PathPatternPrefix::PathSearch),
        _ => fail,
    }
    .parse_next(input)
}

pub fn path_mode_prefix(input: &mut TokenStream) -> ModalResult<Spanned<PathMode>> {
    terminated(path_mode, opt(path_or_paths)).parse_next(input)
}

pub fn path_search_prefix(input: &mut TokenStream) -> ModalResult<Spanned<PathSearchMode>> {
    dispatch! {(peek1, opt(peek1));
        (TokenKind::All, Some(TokenKind::Shortest))
        | (TokenKind::Any, Some(TokenKind::Shortest))
        | (TokenKind::Shortest, _) => shortest_path_search,
        (TokenKind::All, _) => all_path_search,
        (TokenKind::Any, _) => any_path_search,
        _ => fail,
    }
    .parse_next(input)
}

pub fn all_path_search(input: &mut TokenStream) -> ModalResult<Spanned<PathSearchMode>> {
    delimited(TokenKind::All, opt(path_mode), opt(path_or_paths))
        .map(PathSearchMode::All)
        .spanned()
        .parse_next(input)
}

pub fn any_path_search(input: &mut TokenStream) -> ModalResult<Spanned<PathSearchMode>> {
    seq! {PathSearchMode::Any {
        number: opt(number_of_paths),
        mode: opt(path_mode),
        _: opt(path_or_paths),
    }}
    .spanned()
    .parse_next(input)
}

def_parser_alias!(
    number_of_paths,
    non_negative_integer_specification,
    Spanned<NonNegativeInteger>
);
def_parser_alias!(
    number_of_groups,
    non_negative_integer_specification,
    Spanned<NonNegativeInteger>
);

pub fn shortest_path_search(input: &mut TokenStream) -> ModalResult<Spanned<PathSearchMode>> {
    let mut prefix = preceded(
        TokenKind::Shortest,
        terminated((opt(number_of_groups), opt(path_mode)), opt(path_or_paths)),
    );
    let mut suffix = opt(one_of((TokenKind::Group, TokenKind::Groups)).void());
    let mut path_or_group = dispatch! {(prefix.by_ref(), suffix.by_ref());
        ((number, mode), Some(())) => {
            empty.value(PathSearchMode::CountedShortestGroup { number, mode })
        },
        ((Some(number), mode), None) => {
            empty.value(PathSearchMode::CountedShortest { number, mode })
        },
        _ => fail
    }
    .spanned();
    dispatch! {peek2;
        (TokenKind::All, TokenKind::Shortest) => all_shortest_path_search,
        (TokenKind::Any, TokenKind::Shortest) => any_shortest_path_search,
        (TokenKind::Shortest, _) => path_or_group,
        _ => fail
    }
    .parse_next(input)
}

pub fn all_shortest_path_search(input: &mut TokenStream) -> ModalResult<Spanned<PathSearchMode>> {
    delimited(
        (TokenKind::All, TokenKind::Shortest),
        opt(path_mode),
        opt(path_or_paths),
    )
    .map(PathSearchMode::AllShortest)
    .spanned()
    .parse_next(input)
}

pub fn any_shortest_path_search(input: &mut TokenStream) -> ModalResult<Spanned<PathSearchMode>> {
    delimited(
        (TokenKind::Any, TokenKind::Shortest),
        opt(path_mode),
        opt(path_or_paths),
    )
    .map(PathSearchMode::AnyShortest)
    .spanned()
    .parse_next(input)
}

pub fn path_mode(input: &mut TokenStream) -> ModalResult<Spanned<PathMode>> {
    dispatch! {take1;
        TokenKind::Walk => empty.value(PathMode::Walk),
        TokenKind::Trail => empty.value(PathMode::Trail),
        TokenKind::Simple => empty.value(PathMode::Simple),
        TokenKind::Acyclic => empty.value(PathMode::Acyclic),
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn graph_pattern_where_clause(input: &mut TokenStream) -> ModalResult<Spanned<Expr>> {
    preceded(TokenKind::Where, search_condition).parse_next(input)
}

def_parser_alias!(
    parenthesized_path_pattern_where_clause,
    graph_pattern_where_clause,
    Spanned<Expr>
);

pub fn path_variable_declaration(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    terminated(path_variable, TokenKind::Equals).parse_next(input)
}

pub fn subpath_variable_declaration(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    terminated(subpath_variable, TokenKind::Equals).parse_next(input)
}

pub fn order_by_clause(input: &mut TokenStream) -> ModalResult<VecSpanned<SortSpec>> {
    preceded(
        (TokenKind::Order, TokenKind::By),
        separated(1.., sort_specification, TokenKind::Comma),
    )
    .parse_next(input)
}

pub fn sort_specification(input: &mut TokenStream) -> ModalResult<Spanned<SortSpec>> {
    seq! {SortSpec {
        key: sort_key,
        ordering: opt(ordering_specification),
        null_ordering: opt(null_ordering),
    }}
    .spanned()
    .parse_next(input)
}

pub fn ordering_specification(input: &mut TokenStream) -> ModalResult<Spanned<Ordering>> {
    dispatch! {take1;
        TokenKind::Asc | TokenKind::Ascending => empty.value(Ordering::Asc),
        TokenKind::Desc | TokenKind::Descending => empty.value(Ordering::Desc),
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn null_ordering(input: &mut TokenStream) -> ModalResult<Spanned<NullOrdering>> {
    dispatch! {take2;
        (TokenKind::Nulls, TokenKind::First) => empty.value(NullOrdering::First),
        (TokenKind::Nulls, TokenKind::Last) => empty.value(NullOrdering::Last),
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

def_parser_alias!(sort_key, aggregating_value_expression, Spanned<Expr>);

pub fn limit_clause(input: &mut TokenStream) -> ModalResult<Spanned<NonNegativeInteger>> {
    preceded(TokenKind::Limit, non_negative_integer_specification).parse_next(input)
}

pub fn offset_clause(input: &mut TokenStream) -> ModalResult<Spanned<NonNegativeInteger>> {
    preceded(offset_synonym, non_negative_integer_specification).parse_next(input)
}

#[cfg(all(test, feature = "serde"))]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;
    use crate::parser::utils::parse;

    #[test]
    fn test_graph_pattern() {
        let parsed = parse!(graph_pattern, "(a: Person) -[e:Knows]->+ (b: Person)");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_label_expression() {
        let parsed = parse!(label_expression, "!a | b & (c | %)");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_node_pattern_1() {
        let parsed = parse!(
            node_pattern,
            "(a: Person1 & Person2 {id: 123, name: \"Alice\"})"
        );
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_node_pattern_2() {
        let parsed = parse!(node_pattern, "()");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_edge_pattern_1() {
        let parsed = parse!(edge_pattern, "-[e: Knows {since: 2025}]->");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_edge_pattern_2() {
        let parsed = parse!(edge_pattern, "~[]~");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_edge_pattern_3() {
        let parsed = parse!(edge_pattern, "<->");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_pattern_quantifier_1() {
        let parsed = parse!(graph_pattern_quantifier, "*");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_pattern_quantifier_2() {
        let parsed = parse!(graph_pattern_quantifier, "+");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_pattern_quantifier_3() {
        let parsed = parse!(graph_pattern_quantifier, "{123}");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_pattern_quantifier_4() {
        let parsed = parse!(graph_pattern_quantifier, "{,456}");
        assert_yaml_snapshot!(parsed);
    }
}
