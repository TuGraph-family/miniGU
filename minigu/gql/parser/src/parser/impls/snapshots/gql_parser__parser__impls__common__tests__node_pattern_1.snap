---
source: minigu/parser/src/parser/impls/common.rs
expression: parsed
---
- Node:
    variable:
      - a
      - start: 1
        end: 2
    label:
      - Conjunction:
          - - Label: Person1
            - start: 4
              end: 11
          - - Label: Person2
            - start: 14
              end: 21
      - start: 4
        end: 21
    predicate:
      - Property:
          - - name:
                - id
                - start: 23
                  end: 25
              value:
                - Value:
                    Literal:
                      Numeric:
                        Integer:
                          - kind: Decimal
                            integer: "123"
                          - start: 27
                            end: 30
                - start: 27
                  end: 30
            - start: 23
              end: 30
          - - name:
                - name
                - start: 32
                  end: 36
              value:
                - Value:
                    Literal:
                      String:
                        kind: Char
                        literal: Alice
                - start: 38
                  end: 45
            - start: 32
              end: 45
      - start: 22
        end: 46
- start: 0
  end: 47
